src/clientdlg.cpp

```diff
@@ -1193,65 +1193,36 @@ void CClientDlg::OnCLPingTimeWithNumClientsReceived ( CHostAddress InetAddr, int
     ConnectDlg.SetPingTimeAndNumClientsResult ( InetAddr, iPingTime, iNumClients );
 }
 
-void CClientDlg::Connect ( const QString& strSelectedAddress, const QString& strMixerBoardLabel )
+void CClientDlg::OnConnect ( const QString& strMixerBoardLabel )
```

Comment:
Ideally, this signal would get issued only on successful connection.

As we're going cross-thread here, we need Client to supply _any_ listener -- ClientDlg or the Client JSON-RPC -- with full details of the newly established connection.

We should follow this rule for _all_ state changes in the Client -- that should be the _only_ way for ClientDlg and the Client JSON-RPC to acquire knowledge of Client state, no calls to functions across threads.

(The same is true on the server side, too.)

Similarly, to change the state of the Client, the Client should expose slots and ClientDlg or Client JSON-RPC should signal those slots to effect the change in state.  Not call methods across threads.

This does demand a major restructure but I think this could be a good place to start.  Set out _what state changes_ when the Client successfully establishes a connection to a server, create an object instance containing that information, and issue a signal passing that object.  This gives a clearly defined API to that change in state.

Either the message on the signal could be minimal, covering _only_ the state that changes on any connect, or extensive, supplying the _whole_ Client state.  In the latter case, every signal from the Client would pass a new state instance with the full current Client state, making the listener's job much easier.  (I don't like it as much as the minimal approach but it's likely much easier to explain and maintain.  It smacks of "global variables" a bit too much, though...)

Reply:
The global approach could also be wasteful while the small approach could be more difficult. If we restructure correctly, we should aim at the small approach.